webclient/core/migrations/0022_geom_functions.py

```diff
@@ -0,0 +1,205 @@
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
+
+    initial = True
+
+    dependencies = [
+        ('core', '0021_alter_permissions_action'),
+    ]
+
+    operations = [
+        migrations.RunSQL(
+            sql="""
+DROP FUNCTION IF EXISTS public.validateline(geometry);
+
+CREATE OR REPLACE FUNCTION public.validateline(
+    eline geometry,
+    epsg integer
+)
+RETURNS integer
+LANGUAGE plpgsql
+AS $function$
+DECLARE
+    LINE_MAX NUMERIC(8,6);
+    bad_index INTEGER;
+BEGIN
+    -- Nastavení prahu podle EPSG
+    IF epsg = 5514 THEN
+        LINE_MAX := 0.11;  -- S-JTSK (v metrech)
+    ELSIF epsg = 4326 THEN
+        LINE_MAX := 0.000001;  -- WGS84 (v stupních)
+    ELSE
+        RAISE EXCEPTION 'Unsupported EPSG: %', epsg;
+    END IF;
+
+    -- Pokud má geometrie méně než 2 body
+    IF ST_NumPoints(eline) < 2 THEN
+        RETURN -1;
+    END IF;
+
+    -- Vyhledání příliš krátkého segmentu pomocí SQL výrazu
+    SELECT i - 1 INTO bad_index
+    FROM generate_series(2, ST_NumPoints(eline)) AS i
+    WHERE ST_Distance(
+              ST_PointN(eline, i),
+              ST_PointN(eline, i - 1)
+          ) < LINE_MAX
+    LIMIT 1;
+
+    RETURN COALESCE(bad_index, 0);
+END;
+$function$;
+            """,
+            reverse_sql="DROP FUNCTION public.validateline;",
+        ),
+        migrations.RunSQL(
+            sql="""
+DROP FUNCTION IF EXISTS public.validategeom(text);
+
+CREATE OR REPLACE FUNCTION public.validategeom(geom_text text, epsg integer)
+ RETURNS text
+ LANGUAGE plpgsql
+AS $function$
+            declare 
+              back text := 'valid';
+              geom_type_text text;
+			  coord_dim int;
+              geom geometry:=null;
+			  val_result int;
+              bbox geometry;
+    		  allowed_types text[] := ARRAY[
```

Comment:
K uložení bych povolil jen 'ST_Point', 'ST_LineString', 'ST_Polygon' a 'ST_MultiPolygon'. Je to trošku něco jiného než na vstupu při importu přes CSV, kde myslím případně bereme jen ty dílčí části u multiprvků. Multipolygon musí zůstat kvůli katastrům.